Step3 - HTTP 웹 서버 리팩토링

phs1116 opened the pull request:
안녕하세요!
일이 많아 늦게나마 구현하여 코드리뷰 요청드립니다.

전체적으로 요구사항 반영 외에 피드백 사항 반영 및 테스트 추가했습니다.
잘 부탁드립니다.

phs1116 committed:
feat 피드백 반영
테스트코드 패키지 변경
커스텀 익셉션 추가
Cookie 메서드 추가
하드코딩 상수처리

phs1116 committed:
feat RequestFactory 메서드 추상화

phs1116 committed:
feat AbstractController 추가 및 doPost, doGet 구현

phs1116 committed:
feat RequestParameters, Cookies 분리

phs1116 committed:
feat RequestParameters 테스트 추가, 예외 테스트 추가

phs1116 committed:
feat RequestResolver 테스트 추가

phs1116 committed:
feat HttpRequestFactory Test 추가

byjo commented on src/main/java/requesthandler/UserListController.java:
```diff
@@ -29,7 +31,7 @@ public void action(HttpRequest httpRequest, HttpResponse httpResponse) {
 
     private boolean isLogined(String logined) {
-        return !"true".equals(logined);
+        return !LOGIN_VALUE.equals(logined);
```
Boolean.parseBoolean()을 사용해보는건 어떨까요? 
개인적으로는 String value 비교보다 더 명확하다고 느껴져요.
```
private boolean isLogined(String logined) {
   return Boolean.parseBoolean(logined);
}
```
어떠신가요?

그리고 지금 안 사실인데 메서드명은 isLogined인데 반대 값을 return 해주고 있네요 😅

phs1116 replied:
헉 ㅋㅋ... 그러네요. 이름에 맞춰서 수정하겠습니다.

byjo commented on src/main/java/requesthandler/UserLoginController.java:
```diff
@@ -2,35 +2,31 @@
         User user = DataBase.findUserById(userId);
-        if (user.matchPassword(password)) {
+        if (nonNull(user) && user.matchPassword(password)) {
             httpResponse.addCookie(new Cookie("logined", "true"));
```
이전 pr에서 피드백 드렸던 내용이긴 한데요. https://github.com/next-step/jwp-was/pull/59#discussion_r310821338

```suggestion
            httpResponse.addCookie(new Cookie(COOKIE_LOGIN_KEY, String.valueOf(true)));
```
위와 같은 형태는 어떨까요? 
이후에 쿠키 값을 사용할 때 true/false 값의 명확성을 보장할 수 있을 것 같아요.

byjo commented on src/main/java/webserver/request/HttpRequestFactory.java:
```diff
@@ -19,31 +19,44 @@
+
         BufferedReader reader = new BufferedReader(new InputStreamReader(in));
+
+        buildRequestLine(builder, reader);
```
HttpRequestFactory가 많이 깔끔해졌네요 👏
다만, private 메서드인 buildXXX() 메서드는 모두 동일한 역할을 하는데 비해 ,
어떤 메서드는 값을 리턴하고 어떤 메서드는 void로 일관된 규칙이 없네요.

각 private 메서드에서는 파싱하여 생성한 객체를 return 하고, 
create에서 builder에 할당 후 build하는게 더 자연스럽고 각 메서드와 builder의 연관도를 낮출 수 있지 않을까요?

byjo commented on src/main/java/webserver/response/HttpResponse.java:
```diff
@@ -31,6 +31,10 @@ public HttpResponse(HttpStatus httpStatus, byte[] body) {
         this.body = body;
     }
```
사용하지 않는 constructor는 삭제해주세요.
특히 public으로 노출되는 constructor면 더 조심해야 하구요.

byjo commented on src/main/java/webserver/request/HttpRequest.java:
```diff
@@ -8,12 +10,14 @@
 public class HttpRequest {
     private RequestLine requestLine;
```
`HttpRequest` 는 요청에 대한 정보를 담고 있는 클래스겠지요?
그런데 요청에 대한 대부분의 정보는 `RequestLine`이 가지고 있습니다.
이 때문에 요청 정보인 `getPath()`, `getHttpMethod()`메서드를 호출 할 때, 항상 RequestLine을 통하게 돼죠.
심지어 버전 정보를 가져오려 할 땐 아래와 같이 접근 하고 있네요.
https://github.com/next-step/jwp-was/blob/51ba49233264ebc3726fe343d84e9f2c318f5894/src/main/java/webserver/response/HttpResponse.java#L62

RequestLine은 단지 HTTP protocol의 스펙일 뿐입니다. 
우리가 설계하는 클래스가 꼭 이를 따라야 할 필요는 없어요.
`HttpRequest` 의 역할에 맞게 구현해주세요.

phs1116 replied:
RequestLine은 파싱을 위한 결과값으로만 사용하고
실제 HttpRequest에서는 스펙에 맞게 내부 값들을 가지도록 수정했습니다.

byjo commented on src/main/java/webserver/response/HttpResponse.java:
```diff
@@ -43,7 +47,7 @@ public ResponseHeaders getResponseHeaders() {
-    public String getResponseHeader(HttpRequest httpRequest) {
+    public String getResponseHeaderString(HttpRequest httpRequest) {
         StringBuilder stringBuilder = new StringBuilder();
         appendStatusHeader(httpRequest, stringBuilder);
```
`appendStatusHeader()` 메서드를 호출하는 시점은 요청 처리 후 응답을 보내기 바로 직전인데, 
이 때까지 `HttpRequest`에 의존을 하는 것이 맞을까요?
해당 메서드에서 필요한 version이나 content type 정보는 그 이전에 요청 처리를 하면서 response에 set 되어야 할 값들이 아닐까 싶습니다.

이 부분이 정리되면 `RequestHandler.response()` 내부도 깔끔해 보일 것 같네요.
(+) 테스트용 로그는 함께 정리해주세요.
https://github.com/next-step/jwp-was/blob/51ba49233264ebc3726fe343d84e9f2c318f5894/src/main/java/webserver/RequestHandler.java#L45-L48

phs1116 replied:
생성 시점에 HttpRequest로 StatusLine을 생성하는데 정보를 받도록 수정했습니다.
또한 OutStream으로 response를 전달하는 책임은 HttpResponse가 갖도록 변경하였습니다.

byjo reviewed (changes requested):
꼼꼼하게 만든 단위 테스트를 기반으로 리팩토링을 잘 하셨습니다. 👏
(물론 2단계에서 이미 구조화가 잘 되어있기도 했구요.)
다만 각 객체의 역할이 모호한 부분이 있어 이 부분을 정리하고 갔으면 합니다.
코멘트 확인 후 반영 부탁드려요. 

설명이 불충분하거나 문의사항이 있으시다면 깃헙 멘션이나 슬랙 DM 주시구요. 
웹 서버 미션의 끝이 보이니 좀 더 힘내세요! 💪💪💪

phs1116 committed:
feat 피드백 사항 반영
httpResponse 생성시 contentType, httpVersion 가지도록 수정,
response에 대한 책임을 HttpResponse가 가지도록 수정
HttpRequest 구조 변경,
HttpRequestFactory 수정

byjo reviewed (approved):
피드백 반영사항 확인했습니다. 
각 객체의 역할이 좀 더 명료해졌네요. 고생 많으셨습니다. 👏
구조화가 잘 되어있어 4단계 세션 추가는 금방 완료하실 수 있어 보여요.
마지막까지 화이팅입니다!
